src/mscorlib/corefx/System/Runtime/InteropServices/SafeHeapHandleCache.cs

```diff
@@ -0,0 +1,97 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Allows limited thread safe reuse of heap buffers to limit memory pressure.
+    /// 
+    /// This cache does not ensure that multiple copies of handles are not released back into the cache.
+    /// </summary>
+    internal sealed class SafeHeapHandleCache : IDisposable
```

Comment:
> So pinning a char/byte[] array is close to free?

Correct, pinning around *synchronous* PInvokes is close to free.

> I would have expected it to create significant grief for the GC when collecting.

The long term pinning (e.g. some patterns of buffer pinning around asynchronous network operations) tend to cause grief to the GC because of it leads to heap fragmentation over time. 

The short term pinning around synchronous PInvokes is fine because of:
- It is not common that the GC actually needs to run while your synchronous PInvoke is executing
- It is only a problem if the GC is compacting - only a fraction of GCs are compacting
- It is only causing a real grief if multiple consecutive compacting GCs are hitting the pin

> Where is the overhead for SafeHandle?

The interop marshalling layer calls interlock increment/decrement on them when they are passed to PInvoke. The overhead of calling interlock increment/decrement is similar to overhead to taking SpinLock.